### DISCUSSION TITLE
Default library view overloads (?) MPD causing it to disconnect

### DISCUSSION THREAD

**Discussion by hjri** [jcorporation/myMPD#700] [Q&A]
This is most likely due to library size being way too big (almost 3TiB with almost 200k songs and 45k albums according to `mpc stats`). Funnily enough neither Cantata nor M.A.L.P. have any problem displaying entire library, but myMPD "hangs" even with limit set to 25 and stops responding to further queries while MPD can still play music and Cantata/M.A.L.P. can still control it.

It seems that at least with limit set to 25 it will be successful somewhat but still show "MPD disconnected" error. This unfortunately happens when you *just* open library view, I have to work around it by making a shortcut to library that does some bogus search which doesn't hang the myMPD

**Comment by jcorporation** [comment]
Please provide debug logs from myMPD and MPD: https://jcorporation.github.io/myMPD/debug

**Reply by hjri** [reply]
sorry used wrong form initially

It goes something like this:
myMPD:
```
15:58:43 INFO     webserver API request (5): MYMPD_API_DATABASE_ALBUMS_GET // <-- me hitting enter in empty query with limit set to 25 
15:58:43 DEBUG    mympdapi  Leaving mpd idle mode
15:58:43 DEBUG    mympdapi  Handle request
15:58:43 INFO     mympdapi  MYMPD API request (5)(0) MYMPD_API_DATABASE_ALBUMS_GET: {"jsonrpc":"2.0","id":0,"method":"MYMPD_API_DATABASE_ALBUMS_GET","params":{"offset":0,"limit":25,"expression":"","sort":"AlbumArtist","sortdesc":false}}
15:59:08 DEBUG    webserver WS message (4): ping
15:59:33 DEBUG    webserver WS message (4): ping
15:59:42 DEBUG    webserver WS message (2): ping
15:59:58 DEBUG    webserver WS message (4): ping
16:00:23 DEBUG    webserver WS message (4): ping
16:00:37 DEBUG    mympdapi  Push response to web_server_queue for connection 5: {"jsonrpc":"2.0","id":0,"result":{"method":"MYMPD_API_DATABASE_ALBUMS_GET","data":[{"Type":"album","Album":... // bunch of data including file paths etc here
16:00:37 DEBUG    mympdapi  Entering mpd idle mode
16:00:37 ERROR    mympdapi  MPD error: Broken pipe (5)
16:00:37 DEBUG    webserver Got API response for id "5"
16:00:37 DEBUG    webserver Sending response to conn_id 5 (length: 6007): {"jsonrpc":"2.0","id":0,"result":{"method":"MYMPD_API_DATABASE_ALBUMS_GET","data":[{"Type":"album","Album": // again, bunch of data
16:00:37 DEBUG    webserver Sending 6007 bytes to 5
16:00:37 INFO     webserver HTTP request (5): GET /albumart/ //filename redacted
16:00:37 DEBUG    webserver Handle albumart for uri "filename redacted"
16:00:37 DEBUG    webserver No covercache file found
16:00:37 DEBUG    webserver Absolut media_file: //filename redacted
16:00:37 DEBUG    webserver Serving file //filename redacted (image/png)
16:00:37 ERROR    mympdapi  MPD connection failed
16:00:37 DEBUG    mympdapi  Push websocket notify to queue: "{"jsonrpc":"2.0","method":"mpd_disconnected"}"
16:00:37 DEBUG    mympdapi  Trigger event: mympd_disconnected (-5)
16:00:37 INFO     mympdapi  Waiting 2 seconds before reconnection
```

MPD log is a bit more spammy:
first there's a lot of lines like 
```
Mar 02 16:08 : client: [2] command returned 0
Mar 02 16:08 : client: [2] process command "search file "" window 186000:187000"
```
after which there's spam of these:
```
Mar 02 16:08 : client: [2] process command "sticker "list" "song" "path-to-file""
Mar 02 16:08 : client: [2] command returned 0
```

amidst first spam there's client 1 (which seems to be just idling waiting for song change) disconnecting due to timeout:
```
Mar 02 16:07 : client: [1] timeout
Mar 02 16:07 : client: [1] closed
```

Search with 25 limit seems to actually have some result but changing pages takes an obscene amount of time and also causes mpd disconnect.

My understanding is that myMPD lacks its own database/cache and constantly pokes MPD to give it whole fileset which takes a lot of time to aggregate and to transfer which causes myMPD to drop connection(s) due to timeout

**Reply by jcorporation** [reply]
Strange, myMPD creates on startup an album and a sticker cache. This process requests only windows of 1000 songs to avoid reaching protocol events.

After startup MPD is not queried to display the albumgrid.

It seems to be a timeout of the mpd connection. 

15:58:43 DEBUG    mympdapi  Leaving mpd idle mode
16:00:37 DEBUG    mympdapi  Entering mpd idle mode

Setting the timeout greater than 2 minutes should resolv the mpd connection error.

That myMPD needs approximately 2 minutes to query its cache causes this. Can you try current devel version? There are some performance enhancements included.

**Reply by hjri** [reply]
Not sure if I can, since while i was investingating the issue mpd log used about 40MiB of space (it's like 2+ line per each of ~200000 songs in DB + some) and logfile got rotated + don't want to rape my sd card (myMPD/MPD are running on a SBC), same goes for myMPD but at least that is has log in RAM (thanks dietPi) + it leaks literally all the filenames in my collection and I don't want that to be publicly known.

I know that my MPD reports some errors while scanning database, mostly due to (some) (semi-)broken files and unsupported formats or badly encoded filenames.

**Reply by hjri** [reply]
trying larger timeout rn

**Reply by jcorporation** [reply]
I redacted my answer while you answered ;)

**Reply by hjri** [reply]
larger timeout helped but queries still take atrocious amount of time, upgrading from 9.1.0 to 9.1.2 doesn't seem to help, not sure how i can test develop version, i use OBS builds

**Reply by jcorporation** [reply]
OBS does not publish devel builds. You must compile from source. 

9.2.0 will be released very soon.

**Reply by jcorporation** [reply]
I created a dummy database with 200.000 song and 40.000 albums. On my device myMPD returns the list in approximately 10 seconds. The most time is needed for sorting the result utf8 aware and case insensitive, in my test by AlbumArtist. Sorting by LastModified takes only 4 seconds.

**Reply by hjri** [reply]
it's (both MPD and myMPD) running on a (moderately shitty) SBC - Pine64's RockPro. It's naturally slower than full-sized desktop PC. I mean just running aptitude on it is unbearably slow.

**Reply by hjri** [reply]
there are also complications like cue sheets (and virtual folders that are made from them and from... playlists?)

**Reply by hjri** [reply]
Makes me wonder actually, it could be that cantata and malp are fast because they are runing on faster hardware? I wonder if sorting/searching could be made client-side for the myMPD UI (i.e. in JS) to capitalize on potentially faster machine running it.

**Reply by jcorporation** [reply]
I tested with my development laptop with a AMD Ryzen 7 PRO 4750U, that is much faster as your SBC.

Sorting and searching on client side requires to keep the complete albumlist with all metadata e.g. as a json array or an indexeddb database on the client. This can overburden older clients (older smartphones, tablets, etc) and makes the experience even worse.

As far as I know cantata keeps a local copy of the MPD database. I do not know how malp handles this.

**Reply by hjri** [reply]
testing by looking at mpd logs it seems that MALP just caches lists of albums and artists and whatnot, at least it did the `process command "list album group albumartist group date"` when just opening up album list and `process command "list artist"` when opening artists. When actually opening album it does `find album "ALBUM NAME GOES HERE"` and when opening artist it does:
```
list album artist "ARTIST NAME HERE" group albumartist group date
list album albumartist "ARTIST NAME HERE" group albumartist group date
```

I'm guessing that myMPD processes all the songs individually trying to find whatever matches the criteria, which will take long time on a very large dataset (tbh 10 seconds on a decent CPU is also quite bad) while other clients operate on smaller dataset only, i.e. albums/albumartists/artists instead of individual songs.

Interestingly enough, actually searching for something is blazing fast, i.e. if I search `any contains "foo"`, then add `any contains "bar"` and then remove first one search works incredibly fast, the problem is that if there was slow query first (i.e. show all) it slows everything down (and also pegs one cpu core to 100%) and  search essentially doesn't work until that query completes.

Maybe skip some processing/filtering steps and use smaller dataset directly when doing empty queries? It feels like it wastes cycles checking each song if it matches empty criteria or something.

**Reply by jcorporation** [reply]
myMPD also caches only the list of unique albums. The issue is all about sorting. The initial view without filters produces a large dataset that must be sorted. Removing the sort code makes the complete process blazing fast.

Therefore there are 2 ways to fix this issue:
- create a result set that is smaller
- implement a faster sorting algorithm / use a data structure that can be sorted faster

I think building the result set with the window in mind should speed up the process dramatically, because we have only a result set of max 500 to be sorted.

**Reply by jcorporation** [reply]
My first attempt to limit the result set limits only the length of the list to offset + limit. This reduces the response time to under one second. Code is in devel. Can you test?

**Comment by jcorporation** [comment]
myMPD v9.4.0 optimizes sorting in all views and should fix this performance problem finally.
